src/dynamic-launcher.c

```diff
@@ -0,0 +1,868 @@
+{
+  Call *call = call_from_invocation (invocation);
+  g_autoptr(GError) error = NULL;
+  GVariant *launcher_data;
+  g_autofree char *token = NULL;
+  /* same allowed formats as Flatpak */
+  const char *allowed_icon_formats[] = { "png", "jpeg", "svg", NULL };
+  g_autoptr(GVariant) icon_v = NULL;
+
+  if (!xdp_app_info_is_host (call->app_info))
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_NOT_ALLOWED,
+                                             "This call is not available inside the sandbox");
+      return TRUE;
+    }
+
+  if (arg_icon_format == NULL || !g_strv_contains (allowed_icon_formats, arg_icon_format))
+    {
+      g_set_error (&error,
+                   XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
+                   "Dynamic launcher icon format unrecognized (%s)", arg_icon_format);
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  /* Do some validation on the icon before saving it */
+  icon_v = g_variant_get_variant (arg_icon_v);
+  if (!icon_v || !xdp_validate_serialized_icon (icon_v, TRUE /* bytes_only */))
+    {
+      g_set_error (&error,
+                   XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
+                   "Dynamic launcher icon failed validation");
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  launcher_data = g_variant_new ("(svs)", arg_name, icon_v, arg_icon_format);
+  token = g_uuid_string_random ();
+
+  /* Save the token in memory and return it to the caller */
+  set_launcher_data_for_token (token, launcher_data);
+
+  xdp_dynamic_launcher_complete_request_install_token (object, invocation, token);
+
+  return TRUE;
+}
+
+static gboolean
+handle_uninstall (XdpDynamicLauncher    *object,
+                  GDBusMethodInvocation *invocation,
+                  const gchar           *arg_desktop_file_id,
+                  GVariant              *arg_options)
+{
+  Call *call = call_from_invocation (invocation);
+  const char *app_id = xdp_app_info_get_id (call->app_info);
+  g_autoptr(GError) error = NULL;
+  g_autofree char *icon_dir = NULL;
+  g_autofree char *icon_path = NULL;
+  g_autofree char *desktop_dir = NULL;
+  g_autoptr(GFile) icon_file = NULL;
+  g_autoptr(GFile) desktop_file = NULL;
+  g_autoptr(GFile) link_file = NULL;
+  g_autoptr(GKeyFile) desktop_keyfile = NULL;
+
+  if (!validate_desktop_file_id (app_id, arg_desktop_file_id, &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  icon_dir = g_build_filename (g_get_user_data_dir (), "xdg-desktop-portal", "icons", NULL);
+  desktop_dir = g_build_filename (g_get_user_data_dir (), "xdg-desktop-portal", "applications", NULL);
+  g_mkdir_with_parents (desktop_dir, 0700);
+  g_mkdir_with_parents (icon_dir, 0700);
```

Comment:
Feels a bit odd to create directories in the Uninstall() method.

Reply:
Yeah. My thinking was that it would be easier to reason about the error conditions if you only had to worry about the file not existing rather than its parent directories, but I suppose it's `G_IO_ERROR_NOT_FOUND` either way. I can remove these.